### ISSUE TITLE
Error: error updating VM: received an HTTP 500 response - Reason: only root can set 'affinity' config on Proxmox VE 8.1.10

### ISSUE THREAD

**Issue by batonogov** [bpg/terraform-provider-proxmox#1180]
**Describe the bug**
After upgrading to Proxmox VE 8.1.10 I get an error if I modifying **proxmox_virtual_environment_vm**

```sh
Error: error updating VM: received an HTTP 500 response - Reason: only root can set 'affinity' config
```

**To Reproduce**

1. Create a resource **proxmox_virtual_environment_vm**

```tf
resource "proxmox_virtual_environment_vm" "test-01" {
  name        = "test-01"
  description = "Managed by Terraform"
  tags        = ["test"]
  on_boot     = true

  node_name = "pve-02"

  clone {
    vm_id     = "9000"
    node_name = "pve-01"
    retries   = 2
  }

  agent {
    enabled = true
  }

  operating_system {
    type = "l26"
  }

  cpu {
    cores = 2
    type  = "host"
    numa  = true
  }

  memory {
    dedicated = 1024
  }

  disk {
    size         = "20"
    interface    = "virtio0"
    datastore_id = "proxmox-data-02"
    file_format  = "raw"
  }

  network_device {
    bridge = "vmbr0"
    model  = "virtio"
  }

  initialization {
    # interface = "scsi1"
    datastore_id = "proxmox-data-02"
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }
    dns {
      servers = ["77.88.8.8"]
    }
  }
}
```

2. Run

```sh
terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # proxmox_virtual_environment_vm.test-01 will be created
  + resource "proxmox_virtual_environment_vm" "test-01" {
      + acpi                    = true
      + bios                    = "seabios"
      + description             = "Managed by Terraform"
      + id                      = (known after apply)
      + ipv4_addresses          = (known after apply)
      + ipv6_addresses          = (known after apply)
      + keyboard_layout         = "en-us"
      + mac_addresses           = (known after apply)
      + migrate                 = false
      + name                    = "test-01"
      + network_interface_names = (known after apply)
      + node_name               = "pve-02"
      + on_boot                 = true
      + protection              = false
      + reboot                  = false
      + scsi_hardware           = "virtio-scsi-pci"
      + started                 = true
      + stop_on_destroy         = false
      + tablet_device           = true
      + tags                    = [
          + "test",
        ]
      + template                = false
      + timeout_clone           = 1800
      + timeout_create          = 1800
      + timeout_migrate         = 1800
      + timeout_move_disk       = 1800
      + timeout_reboot          = 1800
      + timeout_shutdown_vm     = 1800
      + timeout_start_vm        = 1800
      + timeout_stop_vm         = 300
      + vm_id                   = (known after apply)

      + agent {
          + enabled = true
          + timeout = "15m"
          + trim    = false
          + type    = "virtio"
        }

      + clone {
          + full      = true
          + node_name = "pve-01"
          + retries   = 2
          + vm_id     = 9000
        }

      + cpu {
          + architecture = "x86_64"
          + cores        = 1
          + hotplugged   = 0
          + limit        = 0
          + numa         = true
          + sockets      = 1
          + type         = "host"
          + units        = 1024
        }

      + disk {
          + aio               = "io_uring"
          + backup            = true
          + cache             = "none"
          + datastore_id      = "proxmox-data-02"
          + discard           = "ignore"
          + file_format       = "raw"
          + interface         = "virtio0"
          + iothread          = false
          + path_in_datastore = (known after apply)
          + replicate         = true
          + size              = 20
          + ssd               = false
        }

      + initialization {
          + datastore_id = "proxmox-data-02"

          + dns {
              + servers = [
                  + "77.88.8.8",
                ]
            }

          + ip_config {
              + ipv4 {
                  + address = "dhcp"
                }
            }
        }

      + memory {
          + dedicated      = 1024
          + floating       = 0
          + keep_hugepages = false
          + shared         = 0
        }

      + network_device {
          + bridge      = "vmbr0"
          + enabled     = true
          + firewall    = false
          + mac_address = (known after apply)
          + model       = "virtio"
          + mtu         = 0
          + queues      = 0
          + rate_limit  = 0
          + vlan_id     = 0
        }

      + operating_system {
          + type = "l26"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

proxmox_virtual_environment_vm.test-01: Creating...
proxmox_virtual_environment_vm.test-01: Still creating... [10s elapsed]
proxmox_virtual_environment_vm.test-01: Still creating... [20s elapsed]
proxmox_virtual_environment_vm.test-01: Still creating... [30s elapsed]
proxmox_virtual_environment_vm.test-01: Still creating... [40s elapsed]
proxmox_virtual_environment_vm.test-01: Creation complete after 48s [id=132]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

3. Modify the resource **proxmox_virtual_environment_vm**

```sh
terraform apply
proxmox_virtual_environment_vm.test-01: Refreshing state... [id=132]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # proxmox_virtual_environment_vm.test-01 will be updated in-place
  ~ resource "proxmox_virtual_environment_vm" "test-01" {
        id                      = "132"
        name                    = "test-01"
        tags                    = [
            "test",
        ]
        # (27 unchanged attributes hidden)

      ~ cpu {
          ~ cores        = 1 -> 2
            # (8 unchanged attributes hidden)
        }

        # (7 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

proxmox_virtual_environment_vm.test-01: Modifying... [id=132]
╷
│ Error: error updating VM: received an HTTP 500 response - Reason: only root can set 'affinity' config
│ 
│   with proxmox_virtual_environment_vm.test-01,
│   on xray-02.tf line 1, in resource "proxmox_virtual_environment_vm" "test-01":
│    1: resource "proxmox_virtual_environment_vm" "test-01" {
│ 
╵
```

Please also provide a minimal Terraform configuration that reproduces the issue.

```hcl
# This file is maintained automatically by "terraform init".
# Manual edits may be lost in future updates.

provider "registry.terraform.io/bpg/proxmox" {
  version     = "0.51.0"
  constraints = ">= 0.45.0"
  hashes = [
    "h1:R0TqCGo8dxYJMsUvD8TYfYK5PdOJKkdB7bhdS12Av/U=",
  ]
}
```

**Expected behavior**
I expect the VM resources to change, the VM will restart with the new resources

**Additional context**

- Clustered Proxmox:

```sh
pvesh get /cluster/status
┌─────────────┬─────────────┬─────────┬─────────────┬───────┬───────┬────────┬───────┬────────┬─────────┬─────────┐
│ id          │ name        │ type    │ ip          │ level │ local │ nodeid │ nodes │ online │ quorate │ version │
╞═════════════╪═════════════╪═════════╪═════════════╪═══════╪═══════╪════════╪═══════╪════════╪═════════╪═════════╡
│ cluster     │ pve-cluster │ cluster │             │       │       │        │     3 │        │ 1       │       5 │
├─────────────┼─────────────┼─────────┼─────────────┼───────┼───────┼────────┼───────┼────────┼─────────┼─────────┤
│ node/pve-01 │ pve-01      │ node    │ *.*.*.116   │       │ 1     │      1 │       │ 1      │         │         │
├─────────────┼─────────────┼─────────┼─────────────┼───────┼───────┼────────┼───────┼────────┼─────────┼─────────┤
│ node/pve-02 │ pve-02      │ node    │ *.*.*.117   │       │ 0     │      2 │       │ 1      │         │         │
├─────────────┼─────────────┼─────────┼─────────────┼───────┼───────┼────────┼───────┼────────┼─────────┼─────────┤
│ node/pve-03 │ pve-03      │ node    │ *.*.*.118   │       │ 0     │      3 │       │ 1      │         │         │
└─────────────┴─────────────┴─────────┴─────────────┴───────┴───────┴────────┴───────┴────────┴─────────┴─────────┘
```

- Provider version (ideally it should be the latest version):

```hcl
provider "registry.terraform.io/bpg/proxmox" {
  version     = "0.51.0"
```

- Terraform version:

```sh
Terraform v1.5.7
on darwin_arm64
```

- OS (where you run Terraform from):

```sh
OS: macOS 14.4.1 23E224 arm64
```

- Debug logs (`TF_LOG=DEBUG terraform apply`):

```sh
proxmox_virtual_environment_vm.test-01: Modifying... [id=132]
2024-04-04T12:07:03.213+0300 [INFO]  Starting apply for proxmox_virtual_environment_vm.test-01
2024-04-04T12:07:03.213+0300 [DEBUG] proxmox_virtual_environment_vm.test-01: applying the planned Update change
2024-04-04T12:07:03.217+0300 [DEBUG] provider.terraform-provider-proxmox_v0.51.0: Sending HTTP Request: tf_http_req_body= tf_resource_type=proxmox_virtual_environment_vm tf_rpc=ApplyResourceChange @module=proxmox Accept-Encoding=gzip tf_req_id=470c0d80-b7ac-48c4-0795-178b96460abf Accept=application/json tf_http_op_type=request tf_mux_provider=tf5to6server.v5tov6Server User-Agent=Go-http-client/1.1 tf_http_req_uri=/api2/json/nodes/pve-02/qemu/132/config tf_http_req_version=HTTP/1.1 @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/logging/logging_http_transport.go:162 Authorization=PVEAPIToken=***@pve!terraform=********-****-****-****-************ tf_http_trans_id=5dbb0103-706d-2f8c-472e-6a16d9812b54 tf_provider_addr=registry.terraform.io/bpg/proxmox Host=*.*.*.116:8006 tf_http_req_method=GET timestamp=2024-04-04T12:07:03.217+0300
2024-04-04T12:07:03.267+0300 [DEBUG] provider.terraform-provider-proxmox_v0.51.0: Received HTTP Response: Pragma=no-cache tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=470c0d80-b7ac-48c4-0795-178b96460abf tf_rpc=ApplyResourceChange @module=proxmox Content-Length=761 Expires="Thu, 04 Apr 2024 09:07:03 GMT" tf_http_res_body="{"data":{"digest":"98ce31e8d7d66a8f72553b26f90edd8aa5bcc9f4","balloon":0,"serial0":"socket","cpuunits":1024,"nameserver":"77.88.8.8","ostype":"l26","name":"test-01","description":"Managed by Terraform","numa":1,"meta":"creation-qemu=8.1.2,ctime=1700913616","smbios1":"uuid=17cf01cc-5449-4fca-9f93-5467838a98b8","net0":"virtio=BC:24:11:42:A1:05,bridge=vmbr0,firewall=0","onboot":1,"vga":"memory=16,type=std","ide2":"proxmox-data-02:vm-132-cloudinit,media=cdrom","vmgenid":"2c2e92e2-8cbd-4a9c-8371-acf3744a3508","agent":"enabled=1,fstrim_cloned_disks=0,type=virtio","virtio0":"proxmox-data-02:vm-132-disk-0,size=20G","ipconfig0":"ip=dhcp","memory":"1024","sockets":1,"cpu":"cputype=host","cores":1,"boot":"order=virtio0","scsihw":"virtio-scsi-pci","tags":"test"}}" tf_http_res_status_code=200 tf_http_res_status_reason="200 OK" tf_http_res_version=HTTP/1.1 Connection=[Keep-Alive, Keep-Alive] tf_http_trans_id=5dbb0103-706d-2f8c-472e-6a16d9812b54 Server=pve-api-daemon/3.0 Cache-Control=max-age=0 Content-Type=application/json;charset=UTF-8 Date="Thu, 04 Apr 2024 09:07:03 GMT" tf_http_op_type=response tf_resource_type=proxmox_virtual_environment_vm @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/logging/logging_http_transport.go:162 timestamp=2024-04-04T12:07:03.267+0300
2024-04-04T12:07:03.269+0300 [DEBUG] provider.terraform-provider-proxmox_v0.51.0: Sending HTTP Request: Content-Type=application/x-www-form-urlencoded User-Agent=Go-http-client/1.1 tf_http_op_type=request tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=470c0d80-b7ac-48c4-0795-178b96460abf tf_mux_provider=tf5to6server.v5tov6Server @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/logging/logging_http_transport.go:162 Accept=application/json Authorization=PVEAPIToken=***@pve!terraform=********-****-****-****-************ tf_http_req_body=cores=2&cpu=cputype%3Dhost&cpuunits=1024&delete=affinity%2Cvcpus%2Ccpulimit&name=test-01&numa=1&sockets=1 tf_http_req_method=PUT tf_http_req_uri=/api2/json/nodes/pve-02/qemu/132/config @module=proxmox Host=*.*.*.116:8006 tf_http_req_version=HTTP/1.1 tf_http_trans_id=82d6155d-0dbe-6e80-068d-35073b62d22a tf_resource_type=proxmox_virtual_environment_vm tf_rpc=ApplyResourceChange Accept-Encoding=gzip Content-Length=105 timestamp=2024-04-04T12:07:03.269+0300
2024-04-04T12:07:03.293+0300 [DEBUG] provider.terraform-provider-proxmox_v0.51.0: Received HTTP Response: Content-Type=application/json;charset=UTF-8 Pragma=no-cache tf_provider_addr=registry.terraform.io/bpg/proxmox Cache-Control=max-age=0 Content-Length=13 Date="Thu, 04 Apr 2024 09:07:03 GMT" Expires="Thu, 04 Apr 2024 09:07:03 GMT" Server=pve-api-daemon/3.0 tf_http_op_type=response tf_req_id=470c0d80-b7ac-48c4-0795-178b96460abf @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/logging/logging_http_transport.go:162 @module=proxmox tf_http_res_status_reason="500 only root can set 'affinity' config" tf_http_trans_id=82d6155d-0dbe-6e80-068d-35073b62d22a tf_mux_provider=tf5to6server.v5tov6Server tf_resource_type=proxmox_virtual_environment_vm tf_http_res_body={"data":null} tf_http_res_status_code=500 tf_http_res_version=HTTP/1.1 tf_rpc=ApplyResourceChange timestamp=2024-04-04T12:07:03.293+0300
2024-04-04T12:07:03.294+0300 [ERROR] provider.terraform-provider-proxmox_v0.51.0: Response contains error diagnostic: tf_proto_version=6.4 tf_provider_addr=registry.terraform.io/bpg/proxmox @module=sdk.proto tf_resource_type=proxmox_virtual_environment_vm tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov6/internal/diag/diagnostics.go:58 diagnostic_severity=ERROR diagnostic_summary="error updating VM: received an HTTP 500 response - Reason: only root can set 'affinity' config" tf_req_id=470c0d80-b7ac-48c4-0795-178b96460abf diagnostic_detail= timestamp=2024-04-04T12:07:03.294+0300
2024-04-04T12:07:03.296+0300 [DEBUG] State storage *remote.State declined to persist a state snapshot
2024-04-04T12:07:03.296+0300 [ERROR] vertex "proxmox_virtual_environment_vm.test-01" error: error updating VM: received an HTTP 500 response - Reason: only root can set 'affinity' config
2024-04-04T12:07:03.296+0300 [DEBUG] states/remote: state read serial is: 30; serial is: 30
2024-04-04T12:07:03.296+0300 [DEBUG] states/remote: state read lineage is: 880e059b-458c-bae8-4ccd-38409b97c8ec; lineage is: 880e059b-458c-bae8-4ccd-38409b97c8ec
╷
│ Error: error updating VM: received an HTTP 500 response - Reason: only root can set 'affinity' config
│ 
│   with proxmox_virtual_environment_vm.test-01,
│   on xray-02.tf line 1, in resource "proxmox_virtual_environment_vm" "test-01":
│    1: resource "proxmox_virtual_environment_vm" "test-01" {
│ 
╵
2024-04-04T12:07:03.299+0300 [DEBUG] DELETE https://***/api/v4/projects/***/terraform/state/***/lock
2024-04-04T12:07:03.378+0300 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-04-04T12:07:03.379+0300 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/bpg/proxmox/0.51.0/darwin_arm64/terraform-provider-proxmox_v0.51.0 pid=85483
2024-04-04T12:07:03.379+0300 [DEBUG] provider: plugin exited
```

**Comment by woloss** [comment]
I have the same, but `only root can set 'hugepages' config`. PVE 8.1.4, terraform version 1.3.7 & 1.7.5.

Tried to revert back to v0.48.1 and v0.50.0 (right before affinity and hugepages update), terraform crashing due to updated state.

P.S. test repo had no issues with 0.51.0 and still don't

**Comment by bpg** [comment]
That's probably a regression after #1153, so v0.50.0 should work.

**Comment by bpg** [comment]
@woloss how it crashes? Could you post an error message here?

**Comment by woloss** [comment]
@bpg False alarm, previous runs I had via Atlantis (latest terraform), tried locally and 0.50.0 worked flawlessly. Tried atlantis with terraform v1.3.7 worked fine too, looks like some memory issue with S3 backend plugin, nothing to do with provider.

**Comment by mlinares1998** [comment]
Hi @bpg , came here after having the same problem. 
In my case i tried to change the CPU core count and RAM amount on some VM's got the affinity and hugepages errors respectively.

Downgrading to v0.50.0 indeed fixes the issue for both cases...

PVE 8.1.10 OpenTofu 1.6.2

**Comment by bpg** [comment]
Yeah... the new version automatically adds `delete=affinity` to the API call, even if no affinity was previously defined in the plan.

@batonogov Thanks bunch for the detailed report, it helped me to quickly identify the problem. Will be fixed in v0.50.1

**Comment by bpg** [comment]
@all-contributors please add @batonogov for bug

**Comment by mlinares1998** [comment]
Hi! @bpg! 
Sorry for reopening again, i have tested v0.51.1 and the CPU affinity bug seems to be fixed now.
Unfortunately i'm still experiencing the HTTP 500 error related to hugepages when i try to change a VM allocated RAM...

```
module.pve_virtual_machines["services"].proxmox_virtual_environment_vm.vm: Modifying... [id=112]
╷
│ Error: error updating VM: received an HTTP 500 response - Reason: only root can set 'hugepages' config
│ 
│   with module.pve_virtual_machines["services"].proxmox_virtual_environment_vm.vm,
│   on ../../../../../../modules/pve/qemu/main.tf line 13, in resource "proxmox_virtual_environment_vm" "vm":
│   13: resource "proxmox_virtual_environment_vm" "vm" {
│ 
╵
ERRO[0006] tofu invocation failed in /home/mlinares/easybyte/terraform/env/ops/infra/.terragrunt-cache/SlpvMVecbTivGbI6svv3y0tbtmk/2NGM4drkrPaB1Vhrlkr3XhFZ-gI  prefix=[/home/mlinares/easybyte/terraform/env/ops/infra] 
ERRO[0006] 1 error occurred:
        * [/home/mlinares/easybyte/terraform/env/ops/infra/.terragrunt-cache/SlpvMVecbTivGbI6svv3y0tbtmk/2NGM4drkrPaB1Vhrlkr3XhFZ-gI] exit status 1
```

Should i open a separate ticket for this @bpg ?

**Comment by bpg** [comment]
Ah... thanks for catching this @mlinares1998 !
The problem is exactly the same 😞. No need for the new ticket.
